### DISCUSSION TITLE
Typescript and React table columns issue

### DISCUSSION THREAD

**Discussion by talgat-ruby** [TanStack/table#2571] [General] [answered]
**Describe the bug (required)**
Simple Table with dummy data. But typescript complaining about columns. But why?

**Provide an example via Codesandbox! (required)**
https://codesandbox.io/s/condescending-raman-ud0l8?file=/src/App.tsx:879-886

**Steps To Reproduce (required)**
The `makeData` simply creates dummy array data.
In `App.tsx` line `44` typescript complains on columns.

<img width="689" alt="Screen Shot 2020-07-24 at 09 17 30" src="https://user-images.githubusercontent.com/17277013/88412566-a4134d80-cd8e-11ea-9f03-4e6870b4630a.png">

**Desktop (please complete the following information):**

- OS: Mac
- Browser: Chrome
- Version: 84

**Comment by JensMeire** [comment, marked as answer]
if you cast it as an array of `Column` it works, like this:
```
import { useTable, Column } from "react-table";
...
 const columns = React.useMemo(
    () => [
      {
        Header: t("Customer"),
        accessor: "customer"
      },
      ....
    ] as Column<StatementData>[],
    [t]
  );
```

Can't tell you why tho :)

**Reply by hkar19** [reply]
basically what you did is you cast the entire array into `Column<StatementData>[]`.

each `accessor` must corresponds to keys in `StatementData`, and type `string` is far to general compared to keys in `StatementData`, giving the error.

your casting says you yourself guarante that the entire array is `Column<StatementData>[]` so typescript will believe you.

**Reply by csmattjohnston** [reply]
Thank you this helped me with my use case

**Comment by talgat-ruby** [comment]
@JensMeire you are right. I think this is because `columns` needs to be specified as `Column<T>[]`. 

And I have changed to this
```
const columns = React.useMemo<Column<StatementData>[]>(
  () => [
    {
      Header: t("Customer"),
      accessor: "customer"
    },
   ...
  ],
  [t]
);
```

Thank you!
